internal/controllers/machineset/machineset_controller.go

```diff
@@ -978,75 +1047,105 @@ func (r *Reconciler) updateStatus(ctx context.Context, cluster *clusterv1.Cluste
-		newStatus.DeepCopyInto(&ms.Status)
+		newStatus.ObservedGeneration = s.machineSet.Generation
+		newStatus.DeepCopyInto(&s.machineSet.Status)
 	}
 	switch {
 	// We are scaling up
 	case newStatus.Replicas < desiredReplicas:
-		conditions.MarkFalse(ms, clusterv1.ResizedCondition, clusterv1.ScalingUpReason, clusterv1.ConditionSeverityWarning, "Scaling up MachineSet to %d replicas (actual %d)", desiredReplicas, newStatus.Replicas)
+		conditions.MarkFalse(s.machineSet, clusterv1.ResizedCondition, clusterv1.ScalingUpReason, clusterv1.ConditionSeverityWarning, "Scaling up MachineSet to %d replicas (actual %d)", desiredReplicas, newStatus.Replicas)
 	// We are scaling down
 	case newStatus.Replicas > desiredReplicas:
-		conditions.MarkFalse(ms, clusterv1.ResizedCondition, clusterv1.ScalingDownReason, clusterv1.ConditionSeverityWarning, "Scaling down MachineSet to %d replicas (actual %d)", desiredReplicas, newStatus.Replicas)
+		conditions.MarkFalse(s.machineSet, clusterv1.ResizedCondition, clusterv1.ScalingDownReason, clusterv1.ConditionSeverityWarning, "Scaling down MachineSet to %d replicas (actual %d)", desiredReplicas, newStatus.Replicas)
 		// This means that there was no error in generating the desired number of machine objects
-		conditions.MarkTrue(ms, clusterv1.MachinesCreatedCondition)
+		conditions.MarkTrue(s.machineSet, clusterv1.MachinesCreatedCondition)
 	default:
 		// Make sure last resize operation is marked as completed.
 		// NOTE: we are checking the number of machines ready so we report resize completed only when the machines
 		// are actually provisioned (vs reporting completed immediately after the last machine object is created). This convention is also used by KCP.
 		if newStatus.ReadyReplicas == newStatus.Replicas {
-			if conditions.IsFalse(ms, clusterv1.ResizedCondition) {
+			if conditions.IsFalse(s.machineSet, clusterv1.ResizedCondition) {
 				log.Info("All the replicas are ready", "replicas", newStatus.ReadyReplicas)
 			}
-			conditions.MarkTrue(ms, clusterv1.ResizedCondition)
+			conditions.MarkTrue(s.machineSet, clusterv1.ResizedCondition)
 		}
 		// This means that there was no error in generating the desired number of machine objects
-		conditions.MarkTrue(ms, clusterv1.MachinesCreatedCondition)
+		conditions.MarkTrue(s.machineSet, clusterv1.MachinesCreatedCondition)
 	}
 
 	// Aggregate the operational state of all the machines; while aggregating we are adding the
 	// source ref (reason@machine/name) so the problem can be easily tracked down to its source machine.
-	conditions.SetAggregate(ms, clusterv1.MachinesReadyCondition, collections.FromMachines(filteredMachines...).ConditionGetters(), conditions.AddSourceRef())
+	conditions.SetAggregate(s.machineSet, clusterv1.MachinesReadyCondition, collections.FromMachines(s.machines...).ConditionGetters(), conditions.AddSourceRef())
 
 	return nil
 }
 
+func shouldRequeueForReplicaCountersRefresh(s *scope) ctrl.Result {
+	var replicas int32
```

Comment:
Our webhook is default the replicas field to 1. Are we sure we want to use 0 as a default here?

(Note: the only way to actually have replicas nil here is if the defaulting webhook was removed)

Hm I guess this case is impossible to hit anyway as at least syncReplicas already returns an error if replicas is nil and then we never call this func. Feel free to just resolve